homeassistant/components/media_player/heos.py

```diff
@@ -0,0 +1,217 @@
+    SUPPORT_VOLUME_MUTE, SUPPORT_VOLUME_SET, SUPPORT_VOLUME_STEP)
+
+from homeassistant.const import (CONF_HOST, CONF_NAME, CONF_PASSWORD,
+                                 CONF_USERNAME, STATE_OFF, STATE_PAUSED,
+                                 STATE_PLAYING, STATE_UNKNOWN)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['aioheos==0.1.5']
+
+DEFAULT_NAME = 'HEOS Player'
+
+SUPPORT_HEOS = SUPPORT_PLAY | SUPPORT_STOP | SUPPORT_PAUSE | \
+        SUPPORT_PLAY_MEDIA | SUPPORT_PREVIOUS_TRACK | SUPPORT_NEXT_TRACK | \
+        SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP | \
+        SUPPORT_SEEK
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_HOST):
+    cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME):
+    cv.string,
+    vol.Optional(CONF_USERNAME):
+    cv.string,
+    vol.Optional(CONF_PASSWORD):
+    cv.string,
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discover_info=None):
+    """Set up the HEOS platform."""
+    host = config.get(CONF_HOST)
+    name = config.get(CONF_NAME)
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+
+    heos = HeosMediaPlayer(hass, host, name, username, password)
+
+    yield from heos.heos.connect(
+        host=host, callback=heos.async_update_ha_state)
+
+    async_add_devices([heos])
+
+
+class HeosMediaPlayer(MediaPlayerDevice):
+    """The HEOS player."""
+
+    def __init__(self, hass, host, name, username, password):
+        """Initialize."""
+        from aioheos import AioHeos
+        if not host:
+            _LOGGER.info('No host provided, will try to discover...')
+        self._hass = hass
+        _verbose = False
+        self.heos = AioHeos(
+            loop=hass.loop,
+            host=host,
+            username=username,
+            password=password,
+            verbose=_verbose)
+        self._name = name
+        self._state = None
+
+    @asyncio.coroutine
+    def async_update(self):
+        """Retrieve latest state."""
+        self.heos.request_play_state()
+        self.heos.request_mute_state()
+        self.heos.request_volume()
+        self.heos.request_now_playing_media()
+        return True
+
+    @property
+    def name(self):
+        """Return the name of the device."""
+        return self._name
+
+    @property
+    def volume_level(self):
+        """Volume level of the device (0..1)."""
+        volume = self.heos.get_volume()
+        return float(volume) / 100
+
+    @property
+    def state(self):
+        """Get state."""
+        self._state = self.heos.get_play_state()
+        if self._state == 'stop':
+            return STATE_OFF
+        if self._state == 'pause':
+            return STATE_PAUSED
+        if self._state == 'play':
+            return STATE_PLAYING
+        return STATE_UNKNOWN
```

Comment:
Return `None`.  This will be converted to `STATE_UNKNOWN`

Reply:
I made the change even though I STATE_UNKNOWN is more obvious than None in my opinion.

Reply:
Yeah I think so too, but `None` has special interpretation in the logic that updates the state machine and is converted to `STATE_UNKNOWN`, so to make sure that gets invoked, we have to pass `None`.